### DISCUSSION TITLE
Question: uniqueness of trackable ids (related to refactoring of trackable logging in context of #2012)

### DISCUSSION THREAD

**Discussion by eddiemuc** [cgeo/cgeo#15136] [Q&A]
Dear @cgeo/team, happy new year.

In context of #2012 I discovered that trackable logging needs refactoring to enable a clean new logging approach. In this context I feel that class "TrackableLog" is currently misused as some kind of "Trackable-light" information (and not to actually store a trackable log, as the name would intend).

I noticed that in contrast to geocaches, trackables have a brand property in addition to their tracking code (I mean the public one).

My Question: is a trackable code unique by itself or is it only unique when bundled with a brand?

**Comment by moving-bits** [comment]
Happy new year!

I do not know of any limitations to the two letter prefix a gc.com trackable starts with (and could not find any such limitation on trackable help page (https://www.geocaching.com/help/index.php?pg=kb.printer.friendly&id=10) or trackable order page (https://www.geocaching.com/account/orders/new)), so I guess we have to assume that we need to take the brand name into account. (IIRC we even had some discussion about this what priority to use when a name clash happens and decided to use gc.com > geokrety.)

**Comment by fm-sys** [comment]
Happy new year!

AFAIK the public trackable code itself is unique, but not the secret tracking code.

**Reply by eddiemuc** [reply]
That would make sense, thanks a lot

**Reply by moving-bits** [reply]
Unique = unique for gc.com, but not necessarily across brands
Same goes for secret tracking code, AFAIK (you can search for a trackable using the secret TB code, and it will open the TB details page)

**Reply by fm-sys** [reply]
Trackables always start with TB, Geokretys with GK. At least I haven't seen something else until now. So why should it not be unique across all platforms?

**Reply by moving-bits** [reply]
"TB" on gc.com is the prefix for the public code, everything else can be prefix of the secret code. 
And we did have gc.com trackables in the past that had their secret code starting with "GK", thus name-clashing with GeoKrety.

**Comment by eddiemuc** [comment]
I have a second question. I noticed that in the inventory the (secret) trackingnumber is displayed for my inventory entries, not the (public) geocode. Is this intentionally or an error? (I am reworking this area right now and could correct it)

**Reply by fm-sys** [reply]
~It used to be the public one in the past. Never noticed it, but it currently looks indeed kinda unexpected.~

**Reply by ztNFny** [reply]
The secret code is the only info printed on the TB, so that's all you have available to identify a TB, e.g. when dropping it. Thus the secret number MUST be shown. I've so far never yet used the public number for anything.
Also on GC.com when logging the secret code is shown in black and bigger text, the public code in small grey text - which points to the importance of secret over public.

**Reply by fm-sys** [reply]
That's true, so probably the secret one really makes more sense at this place.

I'm not using trackables that much and my inventory is most of the time empty. That could explain why it surprised me.

**Reply by eddiemuc** [reply]
That makes a lot of sense, thanks for the answer. I will leave the secret code on display.

(Nevertheless, this does not explain why the secret code is used c:geo-internally as a reference key to identify trackables. Because it is clearly not unique. I sitll don't understand why we went through the difficulties of introducing brand as identifier attribute instead of just using the public code - if the public code is unique, which I currently believe)

**Reply by moving-bits** [reply]
Secret code is required for any TB related action except displaying the TB listing, that's why we should store it whenever we have that info. But only in conjunction with brand identifier.

**Reply by Lineflyer** [reply]
> Secret code is required for any TB related action except displaying the TB listing, that's why we should store it whenever we have that info. But only in conjunction with brand identifier.

It should only be stored as long as the respective TB is in your inventory.
Otherwise accidental grabs/logs might happen on TBs which you e.g. already dropped into a cache but look at again and "not know" what you are doing.
IMHO it makes sense that this info is only kind of transient in c:geo (i.e. when searching for a TB via secret code it will be prefilled for the log).

BTW: As this is my first 2024 posting: Happy New Year to you all!

**Comment by kumy** [comment]
I think "it only unique when bundled with a brand" is right.

**Comment by eddiemuc** [comment]
> Unique = unique for gc.com, but not necessarily across brands Same goes for secret tracking code, AFAIK (you can search for a trackable using the secret TB code, and it will open the TB details page)

I am actually asking for the "public" code only (which is called "geocode" in c:geo's Java code). My trivial observation would be same as @fm-sys  that those are unique across platforms (with unique prefixes). For Geocaches this is good enough for c:geo, it is the sole destinction between platforms in code. If this is not good enough for trackables, there needs to be some good explanation.

Leaving that aside, trackables are handled really weird in code today. Most noteable, trackable-related data seems to be stored a lot of times using the (secret) trackingcode/trackingnumber as a key! It is clear that this secret key is NOT unique across platforms, but why was this ever considered to be a good key for string trackable information???

**Reply by moving-bits** [reply]
I can only guess here - probably GeoKrety got added later. There are other places which may be a indicator for this, eg. `trackable.isActive()` does not work for gc.com TB connector and gives weird results if you try using it.

Using secret tracking number may be ok for gc.com TB, but only in conjunction with brand identifier.

On reading a gc.com cache's inventory, you'll get the public codes only. You can't do anything with those code except display the TB listing. For any TB-related action you'll need the secret key. (Don't know how other TB platforms work with respect to this.)

**Comment by eddiemuc** [comment]
Thanks a lot for all the information. To summarize what I gathered:
* The public trackable code (geocode) is unique. It can be used to identify one trackable worldwide across all platforms (in that sense it has the same properties as the geocode for a Geocache). Unlike the secret code it is also avaibale in all contexts (e.g. even when a user doesn't and can't know the secret code, for example because he is just browsing trackable logs). --> It should thus be used as identifier in e.g. databases or hashmap-keys to store data related to a trackable
* The secret trackingnumber is NOT unique (at least not w/o a brand). Also it is not always available to the user. However, IF it is available (e.g. in inventory and most logging types) it is the more important information --> It should be shown to the user in GUIs solely or at least more prominent than the (public) tracking code

With this in mind I can continue my work. Thanks again!

**Comment by bekuno** [comment]
Happy new year also from my side.

The summary is fine. :)
